debian/contrib/shutdown.8
.TH SHUTDOWN 8 "Oct 10, 2016" "" "GNU/Linux System Adminstrator's manual"
.SH NAME
shutdown, reboot, poweroff \- stop the system
.SH SYNOPSIS
.B /sbin/shutdown
.br
.B /sbin/reboot
.br
.B /sbin/halt
.SH DESCRIPTION
SysV-init compatibility scripts, that just invokes
.BR /sbin/init ,
which are expected by graphical desktop environments.
.SH SEE ALSO
.BR init (8)


debian/changelog
runit (2.1.2-19) unstable; urgency=medium

  * Include /usr/local/{sbin,bin} into PATH for runscripts (Closes: #913876)
  * Build-depend on `debhelper-compat' (obsoletes `debian/compat')
  * Provide `/sbin/poweroff' alias for `/sbin/shutdown' (Closes: #914041)

 -- Dmitry Bogatov <KAction@debian.org>  Mon, 19 Nov 2018 13:48:45 +0000

runit (2.1.2-18) unstable; urgency=medium

  * Improve `update-service' script (Thanks: Lorenzo Puliti)
    + Add --auto/--noauto option to update-service
    + Handle relative symlinks
  * Ensure, that halt(8) and reboot(8) commands work properly
    in single mode (Closes: #912937)
    + Thanks: alecfeldman@disroot.org

 -- Dmitry Bogatov <KAction@debian.org>  Thu, 08 Nov 2018 01:21:32 +0000

runit (2.1.2-17) unstable; urgency=medium

  * Add special case into /etc/runit/1 to handle irregular naming
    of some runscripts.
  * Make bin:runit (providing supervision system) recommend packages,
    containing integration with initialization systems. (Closes: #866652)

 -- Dmitry Bogatov <KAction@debian.org>  Sat, 27 Oct 2018 05:10:42 +0000

runit (2.1.2-16) unstable; urgency=medium

  * Create new session for getty-run script to make sure, that
    job control works.
  * Bump standards version to 4.2.1

 -- Dmitry Bogatov <KAction@gnu.org>  Sat, 06 Oct 2018 13:22:58 +0000

runit (2.1.2-15) unstable; urgency=medium

  * Fix spin lock on systems with poor clock (Closes: #878476)
  * Bring back `update-service`, since it is actually used programmatically
    (Closes: #900393)

 -- Dmitry Bogatov <KAction@gnu.org>  Thu, 31 May 2018 07:58:42 +0300

runit (2.1.2-14) unstable; urgency=medium

  * Rebuild package with dh-runit=2.7.3 (Closes: #899242)
  * Do not install `update-service' script.

 -- Dmitry Bogatov <KAction@gnu.org>  Sun, 27 May 2018 20:11:33 +0300

runit (2.1.2-13) unstable; urgency=medium

  * Disable `chkshgrp' test. It requires either root or user with
    supplementary groups; as such it fails with whalebuilder.
    This test is actually about libc, not runit itself.
  * Halt with RB_POWER_OFF, not RB_HALT in debian/contrib/shutdown.c
    to avoid system to freeze in power-on state on halt. It should be
    noted, that currently runit-init runs in sysvinit-compatibility mode,
    and system shutdown actually handled by /sbin/halt -f, not runit itself
    after execution of /etc/runit/3.

 -- Dmitry Bogatov <KAction@gnu.org>  Sun, 13 May 2018 22:40:08 +0300

runit (2.1.2-12) unstable; urgency=medium

  * Improve 'contrib/1' to support boot in single mode. (Closes: #895914)
    + Thanks: Lorenzo Puliti <lorenzo.ru.g@gmail.com>
  * Simplify code (in particular, debian/rules), generating getty runscripts.
  * Improve getty runscripts to not waste CPU when tty is already busy.
    (Closes: #895904)
     + Thanks: Lorenzo Puliti <lorenzo.ru.g@gmail.com>

 -- Dmitry Bogatov <KAction@gnu.org>  Fri, 20 Apr 2018 14:46:17 +0300

runit (2.1.2-11) unstable; urgency=medium

  * Rename contrib/runscript to mk-runscript, since previous
    name conflicted with bin:minicom (Closes: #895607)
  * Improve contrib/runscript (Closes: #873836)
      + fix typo: unprivilleged -> unprivileged
      + ensure that nested runsv processes are not left unmanaged
    - Thanks: chrysn <chrysn@fsfe.org>
  * Remove debugging code from runit.preinst (Closes: #881357)
  * Update standards version to 4.1.4 (no changes needed)
  * Add dependency on recent version of dh-runit, to fix piuparts
    test of fgetty-run.

 -- Dmitry Bogatov <KAction@gnu.org>  Sat, 14 Apr 2018 10:44:53 +0300

runit (2.1.2-10) unstable; urgency=medium

  [ Mathieu Mirmont ]
  * Make compatibility runlevel(8) print "2", since it is default
    runlevel for Debian.
  * Remove /command, absent on Debian systems, from PATH in
    'debian/contrib/1'
  * Adjust 'debian/contrib/3' to SVDIR=/etc/service, not /service.

  [ Dmitry Bogatov ]
  * Replace DEB_BUILD_OPTIONS with DEB_BUILD_MAINT_OPTIONS, since the former
    is meant for user, rebuilding package and overrides the latter.
      + Thanks: Gianfranco Costamagna <locutusofborg@debian.org>
  * Add 'runscript' program into 'runit' binary package, intended to
    simplify for system administrator management of per-user 'runsvdir'
    processes.
  * Incorporate NMU. Bring back runit-init package.
  * Add compiler flags for large file support.
  * Add support for sysvinit control pipe into shutdown script, solving issue
    with reboot just after runit-init installation. (Closes: #861536)
  * Add dependency on initscripts. While initscripts probably could be
    simplified, right now it is simpler to rely on it to make sure, that
    transition from sysvinit is smooth. On other hand, if both runscript
    and initscript are present, runscript is favored.
  * Write CFLAGS and LDFLAGS into conf-ld/conf-cc. Without it, upstream
    build system strips binaries, wasting precious debug information that
    must be stored in -dbg symbols.
  * Check for executable bit on /run/runit.reboot with stat(1), not
    test(1).  If /run is mounted with 'noexec' option, test(1) gives
    unexpected results.
  * Reduce timeout between SIGTERM and SIGKILL on system shutdown.

    Default 7 seconds is reasonable for well-behaving program, and waiting
    longer does not help with programs, which ignore SIGTERM, like
    startx(1) in some configurations.
  * Fix problem with bash completion script when ls(1) output color control
    sequences.
  * Update Vcs-* fields in debian/control. Now on GitLab instance at
    salsa.debian.org
  * Bump compat version to 11, rendering dh-sysuser unneeded.
  * Drop explicit --parallel from debian/rules, since it is default since
    debhelper-10.
  * Use secure URL when referencing to debian/copyright format.
  * Bump standards version to 4.1.3 (removed priority 'extra')
  * Remove version conflicts on sysvinit, since it is no longer present
    in Archives.
  * Add Replaces field into bin:runit-init, since /sbin/runit is moved back

debian/contrib/shutdown.c
// Copyright 2018 Dmitry Bogatov <KAction@gnu.org>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// This source file compiles to following binaries:
//
//  - /sbin/shutdown
//  - /sbin/halt
//  - /sbin/reboot

// They must work when /sbin/init is runit, but PID1 is not. It happens
// when bin:runit-init replaces bin:systemd-sysv or bin:sysvinit-core.
//
// Process of system rebooting/halting has nothing to do with PID1, it
// is all about reboot(2).
//
// reboot(8)/halt(8), provided by sysvinit, when invoked without -f,
// ask PID1 to terminate all processes. As last step, PID1 invokes
// /etc/rc0.d/K08halt or /etc/rc6.d/K08reboot, which, at their turn,
// invoke reboot(8)/halt(8) with -f option. With that option on,
// reboot(2) is called, causing Linux to reboot/halt.
//
// They are not needed for work in command line, since 'runit' uses
// another, more simple interface to system
// but they have been here

// /sbin/reboot binaries. Strictly speaking, they are not
#include <stdbool.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <sys/reboot.h>
#include <assert.h>
#include <string.h>
#include <signal.h>

#define write2(msg) write(2, msg, sizeof(msg))
typedef enum shutdown_action { ACTION_REBOOT, ACTION_HALT } shutdown_action;

/* Copied and slightly simplified from sysvinit-2.88dsf:/src/initreq.h */
struct sysv_request {
	int	magic;			/* Magic number                 */
	int	cmd;			/* What kind of request         */
	int	runlevel;		/* Runlevel to change to        */
	int	sleeptime;		/* Time between TERM and KILL   */
	char	unused[368];
};
#define SYSV_MAGIC              0x03091969
#define SYSV_CMD_RUNLVL         1
#define SYSV_FIFO               "/run/initctl"

/* Luckily, systemd supports sysvinit pipe interface, so there
 * is no need to write separate function interfacing systemd.
 */
static bool
sysv_shutdown(shutdown_action action)
{
	struct sysv_request request;
	request.magic = SYSV_MAGIC;
	request.cmd = SYSV_CMD_RUNLVL;

	switch (action) {
	case ACTION_REBOOT: request.runlevel = '6'; break;
	case ACTION_HALT:   request.runlevel = '0'; break;
	default: return false;
	};

	int fd = open(SYSV_FIFO, O_WRONLY);
	if (fd < 0) {
		return false;
	}
	bool ok = write(fd, &request, sizeof(request)) == sizeof(request);
	close(fd);
	return ok;
}

static bool
syscall_shutdown(shutdown_action action)
{
	sync();
	switch (action) {
	case ACTION_REBOOT:
		reboot(RB_AUTOBOOT);
		break;
	case ACTION_HALT:
		reboot(RB_POWER_OFF);
		break;
	}
	return false;
}

static bool
runit_shutdown(shutdown_action action)
{
	char* args[] = {
		"/sbin/init", (action == ACTION_HALT) ? "0" : "6", NULL };
	execv("/sbin/init", args);
	return false;
}

/* Primitive form of command line parsing. */
static bool
use_force_p(char **argv)
{
	while (*argv) {
		if (strcmp(*argv, "-f") == 0)
			return true;
		argv++;
	}
	return false;
}

int
main(int argc, char **argv)
{
	shutdown_action action = ACTION_HALT;
	bool use_force = false;
	bool ok;

	if (geteuid() != 0) {
		write2("shutdown: must be superuser.\n");
		return 2;
	}

	if (strstr(argv[0], "reboot")) {
		action = ACTION_REBOOT;
	}
	use_force = use_force_p(argv + 1);

	if (use_force) {
		ok = syscall_shutdown(action);
	} else {
		ok = sysv_shutdown(action)
			|| runit_shutdown(action);
		sleep(15);
		/* System should be already down, but still. */
		syscall_shutdown(action);
	}
	return ok ? 0 : 1;
}
